src/foundry/client-esm/canvas/tokens/ring-data.d.mts

```diff
@@ -1,7 +1,163 @@
 import type DataModel from "../../../common/abstract/data.d.mts";
+import type { DataField, DataSchema } from "../../../common/data/fields.d.mts";
+import type { AnyConstructor, SimpleMerge } from "../../../../utils/index.d.mts";
+import type TokenRing from "./ring.d.mts";
 
-declare class DynamicRingData extends DataModel<any> {}
+import fields = foundry.data.fields;
 
-declare namespace DynamicRingData {}
+/**
+ * Dynamic Ring configuration data model.
+ */
+declare class DynamicRingData extends DataModel<DynamicRingData.Schema> {
+  static override defineSchema(): DynamicRingData.Schema;
+}
+
+declare namespace DynamicRingData {
+  interface Schema extends DataSchema {
+    /** The id of this Token Ring configuration. */
+    id: fields.StringField<{ blank: true }>;
+
+    /** The label of this Token Ring configuration. */
+    label: fields.StringField<{ blank: false }>;
+
+    /** The spritesheet path which provides token ring frames for various sized creatures. */
+    spritesheet: fields.FilePathField<{ categories: ["TEXT"]; required: true }>;
+
+    /**
+     * Registered special effects which can be applied to a token ring.
+     * @defaultValue
+     * ```
+     * {
+     *     RING_PULSE: "TOKEN.RING.EFFECTS.RING_PULSE",
+     *     RING_GRADIENT: "TOKEN.RING.EFFECTS.RING_GRADIENT",
+     *     BKG_WAVE: "TOKEN.RING.EFFECTS.BKG_WAVE",
+     *     INVISIBILITY: "TOKEN.RING.EFFECTS.INVISIBILITY"
+     * }
+     * ```
+     */
+    effects: fields.ObjectField<{ initial: Record<string, string> }, Record<string, string>, Record<string, string>>;
```

Comment:
I think you can be more specific here with `initial`. As in I don't believe it'd break if you gave the complete object. Also why did you feel like you had to give more generics than just the first here?

Reply:
![image](https://github.com/user-attachments/assets/09c0bf0f-8497-4c34-8233-9e560c3f9f06)
This would be why I specified the extra generics

Reply:
From Discord this is to limit the keys to strings.